### PR TITLE
Define the activation behavior of a <portal> element.

### LINKED ISSUE #174: Default click behaviour of portals
It feels like we have an opportunity to make these things useful even without JavaScript.

We could make their default click behaviour to activate. This means if the user clicks on a portal, and the site's JavaScript fails to call activate (eg, it throws an error, fails to load, or hasn't loaded yet), they still get to the site they clicked on.

This makes them behave kinda like links by default.

---

This also gives us an opportunity to provide defaults to shift+click, middle click etc. The complicating factor here is: What happens to a portal if it's opened in a new window?

I guess starting a fresh navigation in a new window is least destructive, but doesn't make use of the pre-loaded page.

Alternatively, we could activate the portal in the new window, and do 'something' with the portal in the page. I'm just not sure what.

### PR DESCRIPTION
**Pull request by jeremyroman**
Resolves #174.

***
<a href="https://pr-preview.s3.amazonaws.com/jeremyroman/portals/pull/223.html" title="Last updated on Jul 7, 2020, 5:21 PM UTC (533c846)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/portals/223/7e745af...jeremyroman:533c846.html" title="Last updated on Jul 7, 2020, 5:21 PM UTC (533c846)">Diff</a>

### CODE REVIEW

**Comment by domenic** [comment]
I think we need a bit more than this:

- Prefer not referencing public JS APIs, since they can be overwritten. Instead we should factor out a shared algorithm.
- We need to update the explainer. In particular:
  - https://github.com/WICG/portals#example references #174 
  - https://github.com/WICG/portals#interactivity should get updates and remove a TODO
  - https://github.com/WICG/portals#accessibility : we should consider making portals expose themselves as links now, instead of as buttons

**Comment by jeremyroman** [comment]
> I think we need a bit more than this:
> 
> * Prefer not referencing public JS APIs, since they can be overwritten. Instead we should factor out a shared algorithm.

Hmm, I didn't think that was so. It refers to the steps associated with a WebIDL operation, not an ECMAScript function (and doesn't use [[Get]] or anything that would retrieve an author-overridden value).

If I were to factor it out, would throwing exceptions and creating promises still be appropriate in that algorithm, or would you want the error handling abstracted too? It'd be unfortunate if this simple change complicated stuff that much.

> * We need to update the explainer. In particular:
>   
>   * https://github.com/WICG/portals#example references #174
>   * https://github.com/WICG/portals#interactivity should get updates and remove a TODO
>   * https://github.com/WICG/portals#accessibility : we should consider making portals expose themselves as links now, instead of as buttons

It's not obvious to me whether/why this should be so. It seems separate from #174 to me.

**Comment by jeremyroman** [comment]
also @a4sriniv do you have an opinion on button/link?

**Comment by domenic** [comment]
> If I were to factor it out, would throwing exceptions and creating promises still be appropriate in that algorithm, or would you want the error handling abstracted too? It'd be unfortunate if this simple change complicated stuff that much.

Actually I think this indicates further how it'd be good to factor it out. As-is, it's not clear what happens if an exception is thrown, or what is done with the returned promise (especially if it's rejected; do we mark it as handled?). It'd be ideal to have an algorithm that was specific to user-initiated activations, which could, e.g., no-op for step 2, and assert instead of throw for steps 4 and 5.

> It's not obvious to me whether/why this should be so. It seems separate from #174 to me.

If you're referring to the button/link distinction only, then I agree we could do that separately. The other explainer updates should be done atomically with this, I think.

**Comment by jeremyroman** [comment]
> > If I were to factor it out, would throwing exceptions and creating promises still be appropriate in that algorithm, or would you want the error handling abstracted too? It'd be unfortunate if this simple change complicated stuff that much.
> 
> Actually I think this indicates further how it'd be good to factor it out. As-is, it's not clear what happens if an exception is thrown, or what is done with the returned promise (especially if it's rejected; do we mark it as handled?). It'd be ideal to have an algorithm that was specific to user-initiated activations, which could, e.g., no-op for step 2, and assert instead of throw for steps 4 and 5.

The bulk of the work is already in the **activate a portal browsing context** algorithm. I can duplicate the `activate(options)` steps; I don't see a natural way to factor them out (without reference to DOMException etc).

I will note that the assertion you mention isn't valid, because activation behavior can still be triggered (e.g. by calling `click()` from script) on elements which are not in a browsing context. I do think that in practice UAs will end up printing more or less the same message as a console warning/error that they would have attached to an exception message, but I concede that fact doesn't need to be in the spec.

Let me see what that looks like.

> > It's not obvious to me whether/why this should be so. It seems separate from #174 to me.
> 
> If you're referring to the button/link distinction only, then I agree we could do that separately. The other explainer updates should be done atomically with this, I think.

Yes, sorry, that is what I meant.

**Comment by a4sriniv** [comment]
> also @a4sriniv do you have an opinion on button/link?

The default activation does satisfy the "navigate on click" requirement for links, but we still can't activate portals into a new tab/window (to mirror opening links in a new tab), and I'm not sure how important that is for links.

File: index.bs [hunk 1]
```diff
@@ -630,6 +630,33 @@ spec:url; type:dfn; text:scheme
...
+
+    1. Let |sourceOrigin| be [=this=]'s [=relevant settings object=]'s
+        [=environment settings object/origin=].
+
+    1. Run the steps to [=activate a portal browsing context|activate=] |portalBrowsingContext|
+        in place of |predecessorBrowsingContext| with |sourceOrigin|.
+
+    <div class="note">
+      This is substantially similar to the steps in {{activate(options)}}, with default options.
+      User agents may wish to display suitable console messages under the same conditions that would
```
Comment (domenic) [inline index.bs:655]:
```suggestion
      User agents might wish to display suitable console messages under the same conditions that would
```

Ideally don't use may/must/should in notes.

File: index.bs [hunk 2]
```diff
@@ -630,6 +630,33 @@ spec:url; type:dfn; text:scheme
     portals-cross-origin-load.sub.html
   </wpt>
 
+  <section algorithm="htmlportalelement-activation-behavior">
+    A <{portal}> element's [=activation behavior=] is to run the following steps:
+
+    1. Let |portalBrowsingContext| be the [=guest browsing context=] of [=this=].
+
+    1. If |portalBrowsingContext| is null, abort these steps.
```
Comment (domenic) [inline index.bs:638]: Style nit these days:

```suggestion
    1. If |portalBrowsingContext| is null, return.
```

File: index.bs [hunk 3]
```diff
@@ -630,6 +630,33 @@ spec:url; type:dfn; text:scheme
...
+
+    1. Let |predecessorBrowsingContext| be the [=document browsing context|browsing context=] of
+        [=this=]'s [=node document=].
+
+    1. If |predecessorBrowsingContext| is null, abort these steps.
```
Comment (domenic) [inline index.bs:643]:
```suggestion
    1. If |predecessorBrowsingContext| is null, return.
```

File: index.bs [hunk 4]
```diff
@@ -630,6 +630,33 @@ spec:url; type:dfn; text:scheme
...
+
+    1. If the [=portal state=] of |predecessorBrowsingContext| is not "`none`", abort these steps.
```
Comment (domenic) [inline index.bs:645]:
```suggestion
    1. If the [=portal state=] of |predecessorBrowsingContext| is not "`none`", return.
```

File: index.bs [hunk 5]
```diff
@@ -630,6 +630,33 @@ spec:url; type:dfn; text:scheme
     portals-cross-origin-load.sub.html
   </wpt>
 
+  <section algorithm="htmlportalelement-activation-behavior">
+    A <{portal}> element's [=activation behavior=] is to run the following steps:
+
+    1. Let |portalBrowsingContext| be the [=guest browsing context=] of [=this=].
```
Comment (domenic) [inline index.bs:636]: We're not in a Web IDL method, so `[=this=]` doesn't exist. Probably it's best to say "A `portal` element _el_'s activation behavior is to run the following steps:" and then use _el_ throughout.

File: index.bs (hunk 1 above)
Comment (jeremyroman) [reply index.bs:655]: Sure, okay. I don't think there's much risk of it being misinterpreted as normative, but "might wish" works fine too. :)

File: index.bs (hunk 5 above)
Comment (jeremyroman) [reply index.bs:636]: Done.

**Comment by jeremyroman** [comment]
PTAL

**Comment by domenic** [comment]
Spec LGTM. You up for the explainer updates too?

**Comment by jeremyroman** [comment]
Man, I thought I'd done that but I apparently only imagined it. Done.

**Pull request merged**